pkg/util/module_service.go

```diff
@@ -0,0 +1,92 @@
+package util
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// This service wraps module service, and adds waiting for dependencies to start before starting,
+// and dependant modules to stop before stopping this module service.
+type moduleService struct {
+	services.Service
+
+	service services.Service
+	name    string
+
+	// startDeps, stopDeps return map of service names to services
+	startDeps, stopDeps func(string) map[string]services.Service
+}
+
+// NewModuleService wraps a module service, and makes sure that dependencies are started/stopped before module service starts or stops.
+// If any dependency fails to start, this service fails as well.
+// On stop, errors from failed dependencies are ignored.
+func NewModuleService(name string, service services.Service, startDeps, stopDeps func(string) map[string]services.Service) services.Service {
+	w := &moduleService{
+		name:      name,
+		service:   service,
+		startDeps: startDeps,
+		stopDeps:  stopDeps,
+	}
+
+	w.Service = services.NewBasicService(w.start, w.run, w.stop)
+	return w
+}
+
+func (w *moduleService) start(serviceContext context.Context) error {
+	// wait until all startDeps are running
+	startDeps := w.startDeps(w.name)
+	for m, s := range startDeps {
+		if s == nil {
+			continue
+		}
+
+		level.Debug(Logger).Log("msg", "module waiting for initialization", "module", w.name, "waiting_for", m)
+
+		err := s.AwaitRunning(serviceContext)
+		if err != nil {
+			return fmt.Errorf("failed to start %v, because it depends on module %v, which has failed: %w", w.name, m, err)
+		}
+	}
+
+	// we don't want to let this service to stop until all dependant services are stopped,
+	// so we use independent context here
+	level.Info(Logger).Log("msg", "initialising", "module", w.name)
+	err := w.service.StartAsync(context.Background())
+	if err != nil {
+		return errors.Wrapf(err, "error starting module: %s", w.name)
+	}
+
+	return w.service.AwaitRunning(serviceContext)
+}
+
+func (w *moduleService) run(serviceContext context.Context) error {
+	// wait until service stops, or context is canceled, whatever happens first.
+	// We don't care about exact error here
+	_ = w.service.AwaitTerminated(serviceContext)
+	return w.service.FailureCase()
+}
+
+func (w *moduleService) stop(_ error) error {
+	// wait until all stopDeps have stopped
+	stopDeps := w.stopDeps(w.name)
+	for _, s := range stopDeps {
```

Comment:
Shouldn't we check if `s == nil` here as well? To me it should be the same of `start()`. In Cortex, it's not an issue because in `module_service_wrapper.go` you filter out `nil` deps, but being a generic utility I would consider safer to add the check here as well.

Am I missing anything?

Reply:
You're right. I think I planned to *remove* nil check from both places originally, but having it is a bit safer.